ComparisonConstraints are allocating string on construction

While investigating our test suite's memory usage, NUnit also came up from some paths where a lot of parsed data was processed through loop that checked against LessThanOrEqualConstraint. If you look how all ComparisonConstraint inheritors are implemented, they all allocate Description when constructed even though I would presume it's not needed unless assertion fails?

https://github.com/nunit/nunit/blob/6f806c61eb84f1fed0b47f9b90f1d1e21ec72e14/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs#L35-L38

I would suggest that these would be either lazy evaluated or evaluated every time the getter is called, like for example in EqualConstraint:

https://github.com/nunit/nunit/blob/6f806c61eb84f1fed0b47f9b90f1d1e21ec72e14/src/NUnitFramework/framework/Constraints/EqualConstraint.cs#L399-L421

Yep, that works, thanks for the asnwer.

I'm now having another problem though, NuGetRestore fails with the following error:
 `error NU1603: nunitlite depends on NETStandard.Library (>= 2.0.2-servicing-26420-0) but NETStandard.Library 2.0.2-servicing-26420-0 was not found. An approximate best match of NETStandard.Library 2.0.2 was resolved.`

--- a/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
@@ -87,6 +87,18 @@ namespace NUnit.Framework.Constraints
             {
                 System.Text.StringBuilder sb = new System.Text.StringBuilder(DisplayMessageBase);
                 sb.Append(MsgUtils.FormatValue(_expected));
+                
+                if (_tolerance != null && !_tolerance.IsUnsetOrDefault)
+                {
+                    sb.Append(" within ");
+                    sb.Append(MsgUtils.FormatValue(_tolerance.Amount));
+
+                    if (_tolerance.Mode == ToleranceMode.Percent)
+                    {
+                        sb.Append(" percent");
+                    }
+                }
+                
                 return sb.ToString();
             }
         }
@@ -147,7 +159,6 @@ namespace NUnit.Framework.Constraints
                 throw new InvalidOperationException("Within modifier may appear only once in a constraint expression");
 
             _tolerance = new Tolerance(amount);
-            Description += " within " + MsgUtils.FormatValue(amount);
             return this;
         }
 
@@ -162,7 +173,6 @@ namespace NUnit.Framework.Constraints
             get
             {
                 _tolerance = _tolerance.Percent;
-                Description += " percent";
                 return this;
             }
         }

--- a/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
@@ -22,6 +22,7 @@
 // ***********************************************************************
 
 using System;
+using System.Text;
 using System.Collections;
 using System.Collections.Generic;
 
@@ -162,7 +163,7 @@ namespace NUnit.Framework.Constraints
             if (comparisonText == null)
                 throw new ArgumentNullException(nameof(comparisonText), "Comparison text can not be null");
             
-            System.Text.StringBuilder sb = new System.Text.StringBuilder(comparisonText);
+            StringBuilder sb = new StringBuilder(comparisonText);
             sb.Append(MsgUtils.FormatValue(_expected));
                 
             if (_tolerance != null && !_tolerance.IsUnsetOrDefault)

--- a/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
@@ -155,7 +155,7 @@ namespace NUnit.Framework.Constraints
         /// Provides standard description of what the constraint tests
         /// based on comparison text.
         /// </summary>
-        /// <param name="comparisonText">Describes, what constraint tests, throws <see cref="ArgumentNullException"/>
+        /// <param name="comparisonText">Describes the comparison being tested, throws <see cref="ArgumentNullException"/>
         /// if null</param>
         /// <exception cref="ArgumentNullException">Is thrown when null passed to a method</exception>
         protected string DefaultDescription(string comparisonText)

--- a/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
@@ -78,37 +78,6 @@ namespace NUnit.Framework.Constraints
         }
 
         /// <summary>
-        /// The Description of what this constraint tests, for
-        /// use in messages and in the ConstraintResult.
-        /// </summary>
-        public override string Description
-        {
-            get
-            {
-                System.Text.StringBuilder sb = new System.Text.StringBuilder(DisplayMessageBase);
-                sb.Append(MsgUtils.FormatValue(_expected));
-                
-                if (_tolerance != null && !_tolerance.IsUnsetOrDefault)
-                {
-                    sb.Append(" within ");
-                    sb.Append(MsgUtils.FormatValue(_tolerance.Amount));
-
-                    if (_tolerance.Mode == ToleranceMode.Percent)
-                    {
-                        sb.Append(" percent");
-                    }
-                }
-                
-                return sb.ToString();
-            }
-        }
-
-        /// <summary>
-        /// Protected field overriden by derived class to provide basis for Description construction
-        /// </summary>
-        protected abstract string DisplayMessageBase { get; }
-
-        /// <summary>
         /// Protected function overridden by derived class to actually perform the comparison
         /// </summary>
         protected abstract bool PerformComparison(ComparisonAdapter comparer, object actual, object expected, Tolerance tolerance);
@@ -177,6 +146,39 @@ namespace NUnit.Framework.Constraints
             }
         }
 
+        #endregion
+        
+        #region Protected Methods
+        
+        /// <summary>
+        /// Provides standard description of what the constraint tests
+        /// based on comparison text.
+        /// </summary>
+        /// <param name="comparisonText">Describes, what constraint tests, throws <see cref="ArgumentNullException"/>
+        /// if null</param>
+        /// <exception cref="ArgumentNullException">Is thrown when null passed to a method</exception>
+        protected string DefaultDescription(string comparisonText)
+        {
+            if (comparisonText == null)
+                throw new ArgumentNullException(nameof(comparisonText), "Comparison text can not be null");
+            
+            System.Text.StringBuilder sb = new System.Text.StringBuilder(comparisonText);
+            sb.Append(MsgUtils.FormatValue(_expected));
+                
+            if (_tolerance != null && !_tolerance.IsUnsetOrDefault)
+            {
+                sb.Append(" within ");
+                sb.Append(MsgUtils.FormatValue(_tolerance.Amount));
+
+                if (_tolerance.Mode == ToleranceMode.Percent)
+                {
+                    sb.Append(" percent");
+                }
+            }
+                
+            return sb.ToString();
+        }
+        
         #endregion
     }
 }

--- a/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/ComparisonConstraint.cs
@@ -78,6 +78,25 @@ namespace NUnit.Framework.Constraints
         }
 
         /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        public override string Description
+        {
+            get
+            {
+                System.Text.StringBuilder sb = new System.Text.StringBuilder(DisplayMessageBase);
+                sb.Append(MsgUtils.FormatValue(_expected));
+                return sb.ToString();
+            }
+        }
+
+        /// <summary>
+        /// Protected field overriden by derived class to provide basis for Description construction
+        /// </summary>
+        protected abstract string DisplayMessageBase { get; }
+
+        /// <summary>
         /// Protected function overridden by derived class to actually perform the comparison
         /// </summary>
         protected abstract bool PerformComparison(ComparisonAdapter comparer, object actual, object expected, Tolerance tolerance);

--- a/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
@@ -28,10 +28,6 @@ namespace NUnit.Framework.Constraints
     /// </summary>
     public class GreaterThanConstraint : ComparisonConstraint
     {
-        /// <summary>
-        /// The Description of what this constraint tests, for
-        /// use in messages and in the ConstraintResult.
-        /// </summary>
         private string _description;
         
         /// <summary>
@@ -48,8 +44,9 @@ namespace NUnit.Framework.Constraints
         {
             get
             {
-                if (System.String.IsNullOrEmpty(_description))
+                if (_description == null)
                     _description = DefaultDescription("greater than ");
+                
                 return _description;
             }
         }

--- a/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
@@ -34,9 +34,14 @@ namespace NUnit.Framework.Constraints
         /// Initializes a new instance of the <see cref="GreaterThanConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
-        public GreaterThanConstraint(object expected) : base(expected)
+        public GreaterThanConstraint(object expected) : base(expected) {}
+
+        /// <summary>
+        /// Description basis
+        /// </summary>
+        protected override string DisplayMessageBase
         {
-            Description = "greater than " + MsgUtils.FormatValue(expected);
+            get { return "greater than "; }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
@@ -29,6 +29,12 @@ namespace NUnit.Framework.Constraints
     public class GreaterThanConstraint : ComparisonConstraint
     {
         /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        private string _description;
+        
+        /// <summary>
         /// Initializes a new instance of the <see cref="GreaterThanConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
@@ -40,7 +46,12 @@ namespace NUnit.Framework.Constraints
         /// </summary>
         public override string Description
         {
-            get { return DefaultDescription("greater than "); }
+            get
+            {
+                if (System.String.IsNullOrEmpty(_description))
+                    _description = DefaultDescription("greater than ");
+                return _description;
+            }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanConstraint.cs
@@ -21,8 +21,6 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 // ***********************************************************************
 
-using System;
-
 namespace NUnit.Framework.Constraints
 {
     /// <summary>
@@ -37,11 +35,12 @@ namespace NUnit.Framework.Constraints
         public GreaterThanConstraint(object expected) : base(expected) {}
 
         /// <summary>
-        /// Description basis
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
         /// </summary>
-        protected override string DisplayMessageBase
+        public override string Description
         {
-            get { return "greater than "; }
+            get { return DefaultDescription("greater than "); }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
@@ -29,6 +29,12 @@ namespace NUnit.Framework.Constraints
     public class GreaterThanOrEqualConstraint : ComparisonConstraint
     {
         /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        private string _description;
+        
+        /// <summary>
         /// Initializes a new instance of the <see cref="GreaterThanOrEqualConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
@@ -40,7 +46,12 @@ namespace NUnit.Framework.Constraints
         /// </summary>
         public override string Description
         {
-            get { return DefaultDescription("greater than or equal to "); }
+            get
+            {
+                if (System.String.IsNullOrEmpty(_description))
+                    _description = DefaultDescription("greater than or equal to ");
+                return _description;
+            }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
@@ -28,10 +28,6 @@ namespace NUnit.Framework.Constraints
     /// </summary>
     public class GreaterThanOrEqualConstraint : ComparisonConstraint
     {
-        /// <summary>
-        /// The Description of what this constraint tests, for
-        /// use in messages and in the ConstraintResult.
-        /// </summary>
         private string _description;
         
         /// <summary>
@@ -48,8 +44,9 @@ namespace NUnit.Framework.Constraints
         {
             get
             {
-                if (System.String.IsNullOrEmpty(_description))
+                if (_description == null)
                     _description = DefaultDescription("greater than or equal to ");
+                
                 return _description;
             }
         }

--- a/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
@@ -21,8 +21,6 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 // ***********************************************************************
 
-using System;
-
 namespace NUnit.Framework.Constraints
 {
     /// <summary>
@@ -37,11 +35,12 @@ namespace NUnit.Framework.Constraints
         public GreaterThanOrEqualConstraint(object expected) : base(expected) {}
 
         /// <summary>
-        /// Description basis
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
         /// </summary>
-        protected override string DisplayMessageBase
+        public override string Description
         {
-            get { return "greater than or equal to "; }
+            get { return DefaultDescription("greater than or equal to "); }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/GreaterThanOrEqualConstraint.cs
@@ -21,6 +21,8 @@
 // WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 // ***********************************************************************
 
+using System;
+
 namespace NUnit.Framework.Constraints
 {
     /// <summary>
@@ -32,9 +34,14 @@ namespace NUnit.Framework.Constraints
         /// Initializes a new instance of the <see cref="GreaterThanOrEqualConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
-        public GreaterThanOrEqualConstraint(object expected) : base(expected)
+        public GreaterThanOrEqualConstraint(object expected) : base(expected) {}
+
+        /// <summary>
+        /// Description basis
+        /// </summary>
+        protected override string DisplayMessageBase
         {
-            Description = "greater than or equal to " + MsgUtils.FormatValue(expected);
+            get { return "greater than or equal to "; }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
@@ -29,6 +29,12 @@ namespace NUnit.Framework.Constraints
     public class LessThanConstraint : ComparisonConstraint
     {
         /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        private string _description;
+        
+        /// <summary>
         /// Initializes a new instance of the <see cref="LessThanConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
@@ -40,7 +46,12 @@ namespace NUnit.Framework.Constraints
         /// </summary>
         public override string Description
         {
-            get { return DefaultDescription("less than "); }
+            get
+            {
+                if (System.String.IsNullOrEmpty(_description))
+                    _description = DefaultDescription("less than ");
+                return _description;
+            }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
@@ -28,10 +28,6 @@ namespace NUnit.Framework.Constraints
     /// </summary>
     public class LessThanConstraint : ComparisonConstraint
     {
-        /// <summary>
-        /// The Description of what this constraint tests, for
-        /// use in messages and in the ConstraintResult.
-        /// </summary>
         private string _description;
         
         /// <summary>
@@ -48,8 +44,9 @@ namespace NUnit.Framework.Constraints
         {
             get
             {
-                if (System.String.IsNullOrEmpty(_description))
+                if (_description == null)
                     _description = DefaultDescription("less than ");
+                
                 return _description;
             }
         }

--- a/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
@@ -32,9 +32,14 @@ namespace NUnit.Framework.Constraints
         /// Initializes a new instance of the <see cref="LessThanConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
-        public LessThanConstraint(object expected) : base(expected)
+        public LessThanConstraint(object expected) : base(expected) {}
+
+        /// <summary>
+        /// Description basis
+        /// </summary>
+        protected override string DisplayMessageBase
         {
-            Description = "less than " + MsgUtils.FormatValue(expected);
+            get { return "less than "; }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanConstraint.cs
@@ -35,11 +35,12 @@ namespace NUnit.Framework.Constraints
         public LessThanConstraint(object expected) : base(expected) {}
 
         /// <summary>
-        /// Description basis
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
         /// </summary>
-        protected override string DisplayMessageBase
+        public override string Description
         {
-            get { return "less than "; }
+            get { return DefaultDescription("less than "); }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
@@ -28,12 +28,8 @@ namespace NUnit.Framework.Constraints
     /// </summary>
     public class LessThanOrEqualConstraint : ComparisonConstraint
     {
-        /// <summary>
-        /// The Description of what this constraint tests, for
-        /// use in messages and in the ConstraintResult.
-        /// </summary>
         private string _description;
-        
+
         /// <summary>
         /// Initializes a new instance of the <see cref="LessThanOrEqualConstraint"/> class.
         /// </summary>
@@ -48,8 +44,9 @@ namespace NUnit.Framework.Constraints
         {
             get
             {
-                if (System.String.IsNullOrEmpty(_description))
+                if (_description == null)
                     _description = DefaultDescription("less than or equal to ");
+                
                 return _description;
             }
         }

--- a/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
@@ -32,9 +32,14 @@ namespace NUnit.Framework.Constraints
         /// Initializes a new instance of the <see cref="LessThanOrEqualConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
-        public LessThanOrEqualConstraint(object expected) : base(expected)
+        public LessThanOrEqualConstraint(object expected) : base(expected) {}
+
+        /// <summary>
+        /// Description basis
+        /// </summary>
+        protected override string DisplayMessageBase
         {
-            Description = "less than or equal to " + MsgUtils.FormatValue(expected);
+            get { return "less than or equal to "; }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
@@ -35,11 +35,12 @@ namespace NUnit.Framework.Constraints
         public LessThanOrEqualConstraint(object expected) : base(expected) {}
 
         /// <summary>
-        /// Description basis
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
         /// </summary>
-        protected override string DisplayMessageBase
+        public override string Description
         {
-            get { return "less than or equal to "; }
+            get { return DefaultDescription("less than or equal to "); }
         }
 
         /// <summary>

--- a/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
+++ b/src/NUnitFramework/framework/Constraints/LessThanOrEqualConstraint.cs
@@ -29,6 +29,12 @@ namespace NUnit.Framework.Constraints
     public class LessThanOrEqualConstraint : ComparisonConstraint
     {
         /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        private string _description;
+        
+        /// <summary>
         /// Initializes a new instance of the <see cref="LessThanOrEqualConstraint"/> class.
         /// </summary>
         /// <param name="expected">The expected value.</param>
@@ -40,7 +46,12 @@ namespace NUnit.Framework.Constraints
         /// </summary>
         public override string Description
         {
-            get { return DefaultDescription("less than or equal to "); }
+            get
+            {
+                if (System.String.IsNullOrEmpty(_description))
+                    _description = DefaultDescription("less than or equal to ");
+                return _description;
+            }
         }
 
         /// <summary>

